From 98830613f7f0b66cb688397832a6fd6a77a9b1f3 Mon Sep 17 00:00:00 2001
From: Ludovic Desroches <ludovic.desroches@atmel.com>
Date: Thu, 24 Jan 2013 15:49:10 +0100
Subject: [PATCH 293/319] ARM: at91: cleanup to remove conflicts between
 pinctrl and the old gpio driver

Don't compile old gpio driver when using device tree because we can have
conflicts with pinctrl for instance at91_gpio_suspend/resume called from pm.c.

The old gpio driver is not compiled if CONFIG_PINCTRL_AT91 is set.

Signed-off-by: Ludovic Desroches <ludovic.desroches@atmel.com>
---
 arch/arm/mach-at91/Makefile      |   10 ++++++----
 arch/arm/mach-at91/at91rm9200.c  |    4 ++++
 arch/arm/mach-at91/at91sam9260.c |    4 ++++
 arch/arm/mach-at91/at91sam9261.c |    4 ++++
 arch/arm/mach-at91/at91sam9263.c |    4 ++++
 arch/arm/mach-at91/at91sam9g45.c |    4 ++++
 arch/arm/mach-at91/at91sam9rl.c  |    4 ++++
 arch/arm/mach-at91/board-dt.c    |    4 ++--
 arch/arm/mach-at91/sama5d3.c     |   12 ++----------
 arch/arm/mach-at91/setup.c       |    6 +++++-
 arch/arm/mach-at91/soc.h         |    4 ++--
 11 files changed, 41 insertions(+), 19 deletions(-)

diff --git a/arch/arm/mach-at91/Makefile b/arch/arm/mach-at91/Makefile
index 577676a..8885c8c 100644
--- a/arch/arm/mach-at91/Makefile
+++ b/arch/arm/mach-at91/Makefile
@@ -2,11 +2,16 @@
 # Makefile for the linux kernel.
 #
 
-obj-y		:= irq.o gpio.o setup.o
+obj-y		:= irq.o setup.o
 obj-m		:=
 obj-n		:=
 obj-		:=
 
+ifneq ($(CONFIG_PINCTRL_AT91),y)
+obj-y	+= gpio.o
+obj-y	+= leds.o
+endif
+
 obj-$(CONFIG_AT91_PMC_UNIT)	+= clock.o
 obj-$(CONFIG_AT91_SAM9_ALT_RESET) += at91sam9_alt_reset.o
 obj-$(CONFIG_AT91_SAM9G45_RESET) += at91sam9g45_reset.o
@@ -95,9 +100,6 @@ obj-$(CONFIG_MACH_AT91SAM_DT) += board-dt.o
 # AT91X40 board-specific support
 obj-$(CONFIG_MACH_AT91EB01)	+= board-eb01.o
 
-# Drivers
-obj-y				+= leds.o
-
 # Power Management
 obj-$(CONFIG_PM)		+= pm.o
 obj-$(CONFIG_AT91_SLOW_CLOCK)	+= pm_slowclock.o
diff --git a/arch/arm/mach-at91/at91rm9200.c b/arch/arm/mach-at91/at91rm9200.c
index a3e4710..8e20531 100644
--- a/arch/arm/mach-at91/at91rm9200.c
+++ b/arch/arm/mach-at91/at91rm9200.c
@@ -251,6 +251,7 @@ static void __init at91rm9200_register_clocks(void)
 	clk_register(&pck3);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -270,6 +271,7 @@ static struct at91_gpio_bank at91rm9200_gpio[] __initdata = {
 		.regbase	= AT91RM9200_BASE_PIOD,
 	}
 };
+#endif
 
 static void at91rm9200_idle(void)
 {
@@ -313,9 +315,11 @@ static void __init at91rm9200_initialize(void)
 			| (1 << AT91RM9200_ID_IRQ4) | (1 << AT91RM9200_ID_IRQ5)
 			| (1 << AT91RM9200_ID_IRQ6);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Initialize GPIO subsystem */
 	at91_gpio_init(at91rm9200_gpio,
 		cpu_is_at91rm9200_bga() ? AT91RM9200_BGA : AT91RM9200_PQFP);
+#endif
 }
 
 
diff --git a/arch/arm/mach-at91/at91sam9260.c b/arch/arm/mach-at91/at91sam9260.c
index 0860212..483c9b4 100644
--- a/arch/arm/mach-at91/at91sam9260.c
+++ b/arch/arm/mach-at91/at91sam9260.c
@@ -284,6 +284,7 @@ static void __init at91sam9260_register_clocks(void)
 	clk_register(&pck1);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -300,6 +301,7 @@ static struct at91_gpio_bank at91sam9260_gpio[] __initdata = {
 		.regbase	= AT91SAM9260_BASE_PIOC,
 	}
 };
+#endif
 
 /* --------------------------------------------------------------------
  *  AT91SAM9260 processor initialization
@@ -348,8 +350,10 @@ static void __init at91sam9260_initialize(void)
 	at91_extern_irq = (1 << AT91SAM9260_ID_IRQ0) | (1 << AT91SAM9260_ID_IRQ1)
 			| (1 << AT91SAM9260_ID_IRQ2);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Register GPIO subsystem */
 	at91_gpio_init(at91sam9260_gpio, 3);
+#endif
 }
 
 /* --------------------------------------------------------------------
diff --git a/arch/arm/mach-at91/at91sam9261.c b/arch/arm/mach-at91/at91sam9261.c
index 4d262f3..959a4f5 100644
--- a/arch/arm/mach-at91/at91sam9261.c
+++ b/arch/arm/mach-at91/at91sam9261.c
@@ -242,6 +242,7 @@ static void __init at91sam9261_register_clocks(void)
 	clk_register(&hck1);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -258,6 +259,7 @@ static struct at91_gpio_bank at91sam9261_gpio[] __initdata = {
 		.regbase	= AT91SAM9261_BASE_PIOC,
 	}
 };
+#endif
 
 /* --------------------------------------------------------------------
  *  AT91SAM9261 processor initialization
@@ -288,8 +290,10 @@ static void __init at91sam9261_initialize(void)
 	at91_extern_irq = (1 << AT91SAM9261_ID_IRQ0) | (1 << AT91SAM9261_ID_IRQ1)
 			| (1 << AT91SAM9261_ID_IRQ2);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Register GPIO subsystem */
 	at91_gpio_init(at91sam9261_gpio, 3);
+#endif
 }
 
 /* --------------------------------------------------------------------
diff --git a/arch/arm/mach-at91/at91sam9263.c b/arch/arm/mach-at91/at91sam9263.c
index 09ae108..74e4d42 100644
--- a/arch/arm/mach-at91/at91sam9263.c
+++ b/arch/arm/mach-at91/at91sam9263.c
@@ -275,6 +275,7 @@ static void __init at91sam9263_register_clocks(void)
 	clk_register(&pck3);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -297,6 +298,7 @@ static struct at91_gpio_bank at91sam9263_gpio[] __initdata = {
 		.regbase	= AT91SAM9263_BASE_PIOE,
 	}
 };
+#endif
 
 /* --------------------------------------------------------------------
  *  AT91SAM9263 processor initialization
@@ -326,8 +328,10 @@ static void __init at91sam9263_initialize(void)
 	arm_pm_restart = at91sam9_alt_restart;
 	at91_extern_irq = (1 << AT91SAM9263_ID_IRQ0) | (1 << AT91SAM9263_ID_IRQ1);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Register GPIO subsystem */
 	at91_gpio_init(at91sam9263_gpio, 5);
+#endif
 }
 
 /* --------------------------------------------------------------------
diff --git a/arch/arm/mach-at91/at91sam9g45.c b/arch/arm/mach-at91/at91sam9g45.c
index 90edfd7..02f718d 100644
--- a/arch/arm/mach-at91/at91sam9g45.c
+++ b/arch/arm/mach-at91/at91sam9g45.c
@@ -321,6 +321,7 @@ static void __init at91sam9g45_register_clocks(void)
 	clk_register(&pck1);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -343,6 +344,7 @@ static struct at91_gpio_bank at91sam9g45_gpio[] __initdata = {
 		.regbase	= AT91SAM9G45_BASE_PIOE,
 	}
 };
+#endif
 
 /* --------------------------------------------------------------------
  *  AT91SAM9G45 processor initialization
@@ -371,8 +373,10 @@ static void __init at91sam9g45_initialize(void)
 	arm_pm_restart = at91sam9g45_restart;
 	at91_extern_irq = (1 << AT91SAM9G45_ID_IRQ0);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Register GPIO subsystem */
 	at91_gpio_init(at91sam9g45_gpio, 5);
+#endif
 }
 
 /* --------------------------------------------------------------------
diff --git a/arch/arm/mach-at91/at91sam9rl.c b/arch/arm/mach-at91/at91sam9rl.c
index cbe72e4..5f4f43b 100644
--- a/arch/arm/mach-at91/at91sam9rl.c
+++ b/arch/arm/mach-at91/at91sam9rl.c
@@ -235,6 +235,7 @@ static void __init at91sam9rl_register_clocks(void)
 	clk_register(&pck1);
 }
 
+#ifndef CONFIG_PINCTRL_AT91
 /* --------------------------------------------------------------------
  *  GPIO
  * -------------------------------------------------------------------- */
@@ -254,6 +255,7 @@ static struct at91_gpio_bank at91sam9rl_gpio[] __initdata = {
 		.regbase	= AT91SAM9RL_BASE_PIOD,
 	}
 };
+#endif
 
 /* --------------------------------------------------------------------
  *  AT91SAM9RL processor initialization
@@ -292,8 +294,10 @@ static void __init at91sam9rl_initialize(void)
 	arm_pm_restart = at91sam9_alt_restart;
 	at91_extern_irq = (1 << AT91SAM9RL_ID_IRQ0);
 
+#ifndef CONFIG_PINCTRL_AT91
 	/* Register GPIO subsystem */
 	at91_gpio_init(at91sam9rl_gpio, 4);
+#endif
 }
 
 /* --------------------------------------------------------------------
diff --git a/arch/arm/mach-at91/board-dt.c b/arch/arm/mach-at91/board-dt.c
index 6167531..f500a53 100644
--- a/arch/arm/mach-at91/board-dt.c
+++ b/arch/arm/mach-at91/board-dt.c
@@ -373,8 +373,8 @@ static void __init at91_dt_device_init(void)
 				__u8 manufacturer[4] = "Inlx";
 				__u8 monitor[14] = "AT043TN24";
 				/* set mXT224 and QT1070 IRQ lines as inputs */
-				at91_set_gpio_input(AT91_PIN_PE31, 1);
-				at91_set_gpio_input(AT91_PIN_PE30, 1);
+				gpio_direction_input(AT91_PIN_PE31);
+				gpio_direction_input(AT91_PIN_PE30);
 				/* set LCD configuration */
 				ek_lcdc_data.smem_len = 480 * 272 * 4;
 				memcpy(ek_lcdc_data.default_monspecs->manufacturer, manufacturer, 4);
diff --git a/arch/arm/mach-at91/sama5d3.c b/arch/arm/mach-at91/sama5d3.c
index f726c38..075edb6 100644
--- a/arch/arm/mach-at91/sama5d3.c
+++ b/arch/arm/mach-at91/sama5d3.c
@@ -366,15 +366,7 @@ static void __init sama5d3_map_io(void)
 	init_consistent_dma_size(SZ_8M);
 }
 
-void __init sama5d3_initialize(void)
-{
-	/* Register GPIO subsystem (using DT) */
-	at91_gpio_init(NULL, 0);
-}
-
-struct at91_init_soc __initdata sama5d3_soc __used = {
-	.builtin = 1,
+AT91_SOC_START(sama5d3)
 	.map_io = sama5d3_map_io,
 	.register_clocks = sama5d3_register_clocks,
-	.init = sama5d3_initialize,
-};
+AT91_SOC_END
diff --git a/arch/arm/mach-at91/setup.c b/arch/arm/mach-at91/setup.c
index 2f69ade..a33d4b5 100644
--- a/arch/arm/mach-at91/setup.c
+++ b/arch/arm/mach-at91/setup.c
@@ -40,6 +40,7 @@ void __init at91rm9200_set_type(int type)
 		at91_get_soc_subtype(&at91_soc_initdata));
 }
 
+#ifndef CONFIG_ARCH_AT91_NONE
 void __init at91_init_irq_default(void)
 {
 	at91_init_interrupts(at91_boot_soc.default_irq_priority);
@@ -53,6 +54,9 @@ void __init at91_init_interrupts(unsigned int *priority)
 	/* Enable GPIO interrupts */
 	at91_gpio_irq_setup();
 }
+#else
+#define at91_init_irq_default NULL
+#endif
 
 void __iomem *at91_ramc_base[2];
 EXPORT_SYMBOL_GPL(at91_ramc_base);
@@ -152,7 +156,7 @@ static void __init soc_detect(u32 dbgu_base)
 
 	case ARCH_ID_SAMA5D3:
 		at91_soc_initdata.type = AT91_SOC_SAMA5D3;
-		at91_boot_soc = sama5d3_soc;
+		at91_boot_soc = at91sama5d3_soc;
 		break;
 	}
 
diff --git a/arch/arm/mach-at91/soc.h b/arch/arm/mach-at91/soc.h
index 8ad39c0..1c4c8ae 100644
--- a/arch/arm/mach-at91/soc.h
+++ b/arch/arm/mach-at91/soc.h
@@ -22,7 +22,7 @@ extern struct at91_init_soc at91sam9g45_soc;
 extern struct at91_init_soc at91sam9rl_soc;
 extern struct at91_init_soc at91sam9x5_soc;
 extern struct at91_init_soc at91sam9n12_soc;
-extern struct at91_init_soc sama5d3_soc;
+extern struct at91_init_soc at91sama5d3_soc;
 
 #define AT91_SOC_START(_name)				\
 struct at91_init_soc __initdata at91##_name##_soc	\
@@ -71,5 +71,5 @@ static inline int at91_soc_is_enabled(void)
 #endif
 
 #if !defined(CONFIG_SOC_SAMA5D3)
-#define sama5d3_soc	at91_boot_soc
+#define at91sama5d3_soc	at91_boot_soc
 #endif
-- 
1.7.10.4

